## **Introduction**

London is a quite popular tourist and vacation destination for people all around the world. It is diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and draw insights to what they look like now.

## **Business Problem**

The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer.

## **Data Description**

We require geolocation data for London. Postal codes of city serves as a starting point. Using Postal codes we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.


## **London**

To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London

This wikipedia page has information about all the neighbourhoods, we limit it London.

1. borough : Name of Neighbourhood
2. town : Name of borough
3. post_code : Postal codes for London

This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

## **ArcGIS API**

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups.

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data.

1. latitude : Latitude for Neighbourhood
2. longitude : Longitude for Neighbourhood

Based on all the information collected for London, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

## **Methodology**

We will be creating our model with the help of Python so we start off by importing all the required packages.

In [ ]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

Package breakdown:

1. Pandas : To collect and manipulate data in JSON and HTMl and then data analysis
2. requests : Handle http requests
3. matplotlib : Detailing the generated maps
4. folium : Generating maps of London
5. sklearn : To import Kmeans which is the machine learning model that we are using.

The approach taken here is to explore the city, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

## **Exploring London**

## Neighborhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

### **Data Collection**

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [ ]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

In [ ]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                       Download coordinates as: KML,
             Location                     London borough  ... Dial code OS grid ref
 0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
 1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
 2          Addington                         Croydon[8]  ...       020    TQ375645
 3         Addiscombe                         Croydon[8]  ...       020    TQ345665
 4        Albany Park                             Bexley  ...       020    TQ478728
 ..               ...                                ...  ...       ...         ...
 526         Woolwich                          Greenwich  ...       020    TQ435795
 527   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
 528  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ2258

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [ ]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough  ... Dial code OS grid ref
0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
2          Addington                         Croydon[8]  ...       020    TQ375645
3         Addiscombe                         Croydon[8]  ...       020    TQ345665
4        Albany Park                             Bexley  ...       020    TQ478728
..               ...                                ...  ...       ...         ...
526         Woolwich                          Greenwich  ...       020    TQ435795
527   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
528  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ225815
529          Yeading                         Hillingdon  ...       020    TQ115825
530         Yiewsley                         Hillingdon  ...       020    TQ063804

[531 rows x 6 columns]

### **Data Pre-Processing**

In [ ]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough  ... Dial code OS_grid_ref
0         Abbey Wood              Bexley, Greenwich [7]  ...       020    TQ465785
1              Acton  Ealing, Hammersmith and Fulham[8]  ...       020    TQ205805
2          Addington                         Croydon[8]  ...       020    TQ375645
3         Addiscombe                         Croydon[8]  ...       020    TQ345665
4        Albany Park                             Bexley  ...       020    TQ478728
..               ...                                ...  ...       ...         ...
526         Woolwich                          Greenwich  ...       020    TQ435795
527   Worcester Park       Sutton, Kingston upon Thames  ...       020    TQ225655
528  Wormwood Scrubs             Hammersmith and Fulham  ...       020    TQ225815
529          Yeading                         Hillingdon  ...       020    TQ115825
530         Yiewsley                         Hillingdon  ...       020    TQ063804

[531 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### **Feature Selection**

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [ ]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [ ]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
526                          Greenwich          LONDON       SE18
527       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528             Hammersmith and Fulham          LONDON        W12
529                         Hillingdon           HAYES        UB4
530                         Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [ ]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
526                       Greenwich          LONDON       SE18
527    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528          Hammersmith and Fulham          LONDON        W12
529                      Hillingdon           HAYES        UB4
530                      Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

In [ ]:
df1.shape

(531, 3)

We currently have 531 records and 3 columns of our data. It's time to perform Feature Engineering

### **Feature Engineering**

We can only focusing on the neighbourhoods of London, so performing the changes

In [ ]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
521                       Redbridge                  LONDON  IG8, E18
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
525                          Barnet                  LONDON       N12
526                       Greenwich                  LONDON      SE18
528          Hammersmith and Fulham                  LONDON       W12

[308 rows x 3 columns]

In [ ]:
df1.shape

(308, 3)

We now have only 308 rows. We can proceed with our further steps. Getting some descriptive statistics


In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308 entries, 0 to 528
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    308 non-null    object
 1   town       308 non-null    object
 2   post_code  308 non-null    object
dtypes: object(3)
memory usage: 9.6+ KB


## **Gelolocations of the London Neighbourhoods**

### **ArcGis API**

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so.

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [ ]:
pip install arcgis

  Using cached https://files.pythonhosted.org/packages/c5/8f/51e89ce52a085483359217bc72cdbf6e75ee595d5b1d4b5ade40c7e018b8/jsonschema-3.2.0-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/77/e8/c3cf72a32a697256608d5fa96360c431adec6e1c6709ba7f13f99ff5ee04/jupyter_client-6.1.12-py3-none-any.whl
ERROR: google-colab 1.0.0 has requirement tornado~=5.1.0; python_version >= "3.0", but you'll have tornado 6.1 which is incompatible.
  Found existing installation: jsonschema 2.6.0
    Uninstalling jsonschema-2.6.0:
      Successfully uninstalled jsonschema-2.6.0
  Found existing installation: jupyter-client 5.3.5
    Uninstalling jupyter-client-5.3.5:
      Successfully uninstalled jupyter-client-5.3.5


In [ ]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()


Defining London arcgis geocode function to return latitude and longitude

In [ ]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)


Checking sample data

In [ ]:
c = get_x_y_uk('SE2')
c

'51.499741450000045,0.12406135200006929'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [ ]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
521    IG8, E18
522         IG8
525         N12
526        SE18
528         W12
Name: post_code, Length: 308, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [ ]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.499741450000045,0.12406135200006929
1        51.49776500000007,-0.2558519459999502
6      51.513145000000065,-0.07873298599997725
7       51.51462500000008,-0.11486033199997792
9        51.48249000000004,0.11919361600007505
                        ...                   
521     51.51180000000005,-0.07129021599996577
522    51.507408360000056,-0.12769869299995662
525     51.45271983530966,-0.33010122381629187
526     51.50312900000006,-0.10802518599996347
528    51.515085000000056,-0.24269643599996016
Name: post_code, Length: 308, dtype: object

### **Latitude**

Extracting the latitude from our previously collected coordinates

In [ ]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.499741450000045
1       51.49776500000007
6      51.513145000000065
7       51.51462500000008
9       51.48249000000004
              ...        
521     51.51180000000005
522    51.507408360000056
525     51.45271983530966
526     51.50312900000006
528    51.515085000000056
Name: post_code, Length: 308, dtype: object

### **Longitude**

In [ ]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12406135200006929
1       -0.2558519459999502
6      -0.07873298599997725
7      -0.11486033199997792
9       0.11919361600007505
               ...         
521    -0.07129021599996577
522    -0.12769869299995662
525    -0.33010122381629187
526    -0.10802518599996347
528    -0.24269643599996016
Name: post_code, Length: 308, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage




In [ ]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough  ... longitude
0                Bexley, Greenwich   ...  0.124061
1    Ealing, Hammersmith and Fulham  ... -0.255852
6                              City  ... -0.078733
7                       Westminster  ... -0.114860
9                           Bromley  ...  0.119194
..                              ...  ...       ...
521                       Redbridge  ... -0.071290
522       Redbridge, Waltham Forest  ... -0.127699
525                          Barnet  ... -0.330101
526                       Greenwich  ... -0.108025
528          Hammersmith and Fulham  ... -0.242696

[308 rows x 5 columns]

In [ ]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

### **Co-ordinates for London**


Getting the geocode for London to help visualize it on the map

In [ ]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.12769869299995662

In [ ]:
london_lat_coords

51.507408360000056

## **Visulaize map of London**

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [ ]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### **Venues in London**

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [ ]:
CLIENT_ID = 'XURQSAQYGFHL25ROIW2GUAKPVVOSDOYLKPKP0JK1AWKHUHGB' 
CLIENT_SECRET = '2YA4KWL1VQAS4QGSAPQL5OAJT20TTX1GRBAPVJQ1QIZ0ZFH5'
VERSION = '20210708' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [ ]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [ ]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and Chelsea, Hammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames


Sampling our Data

In [ ]:
venues_in_London.head()

Neighbourhood  ...            Venue Category
0              Bexley, Greenwich   ...                      Lake
1  Ealing, Hammersmith and Fulham  ...  Mediterranean Restaurant
2  Ealing, Hammersmith and Fulham  ...               Coffee Shop
3  Ealing, Hammersmith and Fulham  ...    Argentinian Restaurant
4  Ealing, Hammersmith and Fulham  ...               Coffee Shop

[5 rows x 5 columns]

In [ ]:
venues_in_London.shape

(11659, 5)

Wow, we have scraped together 11659 records for venues. This will definitely make the clustering interesting.

### **Grouping by Venue Categories**

We need to now see how many Venue Categories are there for further processing

In [ ]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  ...                        Venue
Venue Category                          ...                             
Accessories Store    Camden, Islington  ...           James Smith & Sons
Adult Boutique                 Hackney  ...  Sh! Women's Erotic Emporium
African Restaurant         Westminster  ...                 Jungle Braai
American Restaurant        Westminster  ...             Prairie Fire BBQ
Antique Shop                  Lewisham  ...           Lots Road Auctions
...                                ...  ...                          ...
Wine Shop                  Westminster  ...                  Wine Pantry
Winery                         Lambeth  ...             The Wine Parlour
Wings Joint          Camden, Islington  ...                 Chicken Shop
Yoga Studio                Westminster  ...                    yogahaven
Yoshoku Restaurant   Camden, Islington  ...                   Shack-Fuyu

[311 rows x 4 columns]

### **One Hot Encoding**

We need to Encode our venue categories to get a better result for our clustering

In [ ]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Accessories Store  Adult Boutique  ...  Yoga Studio  Yoshoku Restaurant
0                      0               0  ...            0                   0
1                      0               0  ...            0                   0
2                      0               0  ...            0                   0
3                      0               0  ...            0                   0
4                      0               0  ...            0                   0
...                  ...             ...  ...          ...                 ...
11654                  0               0  ...            0                   0
11655                  0               0  ...            0                   0
11656                  0               0  ...            0                   0
11657                  0               0  ...            0                   0
11658                  0               0  ...            0                   0

[11659 rows x 311 columns]

In [ ]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  ...  Yoshoku Restaurant
0              Bexley, Greenwich   ...                   0
1  Ealing, Hammersmith and Fulham  ...                   0
2  Ealing, Hammersmith and Fulham  ...                   0
3  Ealing, Hammersmith and Fulham  ...                   0
4  Ealing, Hammersmith and Fulham  ...                   0

[5 rows x 312 columns]

### Venue Categories Mean Value

We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [ ]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Accessories Store  ...  Yoga Studio  Yoshoku Restaurant
0                 Barnet                0.0  ...          0.0                 0.0
1  Barnet, Brent, Camden                0.0  ...          0.0                 0.0
2                 Bexley                0.0  ...          0.0                 0.0
3      Bexley, Greenwich                0.0  ...          0.0                 0.0
4     Bexley, Greenwich                 0.0  ...          0.0                 0.0

[5 rows x 312 columns]

Let's make a function to get the top most common venue categories

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

Creating a function to label the columns of the venue correctly

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

### **Top venue categories**

In [ ]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood  ... 10th Most Common Venue
0                 Barnet  ...          Grocery Store
1  Barnet, Brent, Camden  ...                   Farm
2                 Bexley  ...            Fabric Shop
3      Bexley, Greenwich  ...     Falafel Restaurant
4     Bexley, Greenwich   ...   Fast Food Restaurant

[5 rows x 11 columns]

### **Model Building**

### **K Means**

In [ ]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

### **Labelling Clustered Data**

In [ ]:
kmeans_london.labels_

array([1, 0, 1, 3, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [ ]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [ ]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough  ...    10th Most Common Venue
0              Bexley, Greenwich   ...      Fast Food Restaurant
1  Ealing, Hammersmith and Fulham  ...  Mediterranean Restaurant
6                            City  ...                  Wine Bar
7                     Westminster  ...        Chinese Restaurant
9                         Bromley  ...               Fabric Shop

[5 rows x 16 columns]

Drop all the NaN values to prevent data skew

In [ ]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

### **Visualizing the clustered neiighbourhood**

In [ ]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

### **Examining our Clusters**

Cluster 1

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
121  LONDON               1  ...    Falafel Restaurant                   Farm

[1 rows x 12 columns]

Cluster 2

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...    10th Most Common Venue
1                    LONDON  ...  Mediterranean Restaurant
6                    LONDON  ...                  Wine Bar
7                    LONDON  ...        Chinese Restaurant
9                    LONDON  ...               Fabric Shop
10                   LONDON  ...               Pizza Place
..                      ...  ...                       ...
521                  LONDON  ...            Sandwich Place
522  LONDON, WOODFORD GREEN  ...                     Plaza
525                  LONDON  ...             Grocery Store
526                  LONDON  ...           Thai Restaurant
528                  LONDON  ...         Indian Restaurant

[303 rows x 12 columns]

Cluster 3

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
0  LONDON               3  ...        Farmers Market   Fast Food Restaurant

[1 rows x 12 columns]

Cluster 4

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  ...  10th Most Common Venue
167  LONDON, WELLING  ...      Falafel Restaurant
457    LONDON, ERITH  ...      Falafel Restaurant

[2 rows x 12 columns]

Cluster 5

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
270  LONDON               5  ...           Fabric Shop     Falafel Restaurant

[1 rows x 12 columns]

## **Results and Discusion**

The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to have a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores.F For leisure, the neighbourhoods are set up to have lots of parks, gyms andart galary.

Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.





## **Conclusion**

The purpose of this project was to explore the city of London and see how attractive it is to potential tourists and migrants. We explored the city based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods city had a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

London seems to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture. Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.